cpu/cortexm_common: add inlined header only def for irq_%

fjmolinas opened the pull request:
### Contribution description

irq_% are not inlined by the compiler which leads to it branching
to a function that actually implement a single machine instruction.

Inlining these functions makes the call more efficient as well as
saving some bytes in ROM.

This could probably be done for more arch, but I wanted to first do
it for a single arch first.

### Testing procedure

- All applications should still work

- check footprint difference, `BUILD_IN_DOCKER=1 BOARD=<board>make -C examples/gnrc_networking`

**iotlab-m3 cortexm3:**

- PR

```
  text    data     bss     dec     hex filename
  87536     184   19240  106960   1a1d0 /data/riotbuild/riotbase/examples/gnrc_networking/bin/iotlab-m3/gnrc_networking.elf
```

- master

```
  text    data     bss     dec     hex filename
  87740     184   19240  107164   1a29c /data/riotbuild/riotbase/examples/gnrc_networking/bin/iotlab-m3/gnrc_networking.elf
```

**samr21-xpro cortexm0:**

- PR

```
  text    data     bss     dec     hex filename
  91488     184   19240  110912   1b140 /data/riotbuild/riotbase/examples/gnrc_networking/bin/samr21-xpro/gnrc_networking.elf
```

- master

```
  text    data     bss     dec     hex filename
  91588     184   19240  111012   1b1a4 /data/riotbuild/riotbase/examples/gnrc_networking/bin/samr21-xpro/gnrc_networking.elf
```

**nrf52dk cortexm4:**

- PR

```
  text    data     bss     dec     hex filename
  162048     284   47616  209948   3341c /data/riotbuild/riotbase/examples/gnrc_networking/bin/nrf52dk/gnrc_networking.elf
```

- master

```
  text    data     bss     dec     hex filename
  162248     284   47616  210148   334e4 /data/riotbuild/riotbase/examples/gnrc_networking/bin/nrf52dk/gnrc_networking.elf
```

### Issues/PRs references

Came up a while ago in #11919

maribu commented:
This is something I have wanted to see for a long time :-)

It is good to do this one arch at a time. I would also like to see the return value to be lets call it `uword_t` or `irq_state_t`, which is 32 bit wide on a 32 bit system, 16 bit on a 16 bit system, and 8 bit wide on a 8 bit platform.

Every platform I am aware of guarantees to disable and restore IRQs in one CPU cycle. As at most a word can be read/written in a cycle, this gives us the word as upper limit for the IRQ state.

The AVR platform would greatly profit if it would be both `inline`d and word sized, as than the call `irq_disable()` would translate to one instruction.

(But changing the return type is something that deserves its own PR. Especially as a lot of callers have to be updated.)

maribu commented on core/include/irq.h:
```diff
@@ -27,6 +27,51 @@
 extern "C" {
 #endif
 
+#ifdef MODULE_CORTEXM_COMMON
```
While I wish that we would transition to an `inline`able implementation (without `LTO=1`) by the end of next week, I doubt we will manage. So maybe it is better to have a more maintainable guard. How about this:

Every platform except CortexM defines `LEGACY_IRQ_API`. And here we check for `#ifndef LEGACY_IRQ_API`. This would also keep core platform independent (as the check is for a feature, rather than a platform).

fjmolinas replied:
Ok I did that defining them in `cpu_conf.h`, le me know if you think there is a better place.

maribu replied:
I would also have put it just there

maribu commented on cpu/cortexm_common/include/irq_arch.h:
```diff
@@ -18,24 +18,33 @@
+static inline __attribute__((always_inline)) __attribute__((used)) unsigned int
+irq_enable(void)
 {
     __enable_irq();
     return __get_PRIMASK();
```
Mind to sneak in an unrelated bugfix as separate commit? It should be:

```C
unsigned result = __get_PRIMASK();
__enable_irq();
return result;
```

(See [API doc](http://api.riot-os.org/group__core__irq.html#gac5652462dfaa82609d3f88f6b09519ae))

fjmolinas replied:
done

fjmolinas commented:
Rebased.

fjmolinas commented:
Will trigger murdock once to see If I missed an arch.

maribu commented:
Code looks good to me. I think you can squash. (Also: The missing `#ifdef __cplusplus` in `cpu/esp_common/include/cpu_conf_common.h` can be just squashed in.)

fjmolinas commented:
> Code looks good to me. I think you can squash. (Also: The missing #ifdef __cplusplus in cpu/esp_common/include/cpu_conf_common.h can be just squashed in.)

Squashed. I triggered tests as well for good measure.

fjmolinas commented:
Missed a `cpu_conf`.

kYc0o commented:
Nice PR! I really hope this improves the current situation.

Please verify how this change affects stm32l152. I remember we had lots of issues regarding this function. As for #11919 I can see that this function was replaced, maybe it's worth to try it again and check if this also solves the problem. Unfortunately I don't have the hardware anymore, so I can't try by myself.

fjmolinas commented:
> Please verify how this change affects stm32l152. I remember we had lots of issues regarding this function. As for #11919 I can see that this function was replaced, maybe it's worth to try it again and check if this also solves the problem. Unfortunately I don't have the hardware anymore, so I can't try by myself.

I have the `BOARD` and I did test it, it still works, I can lunch all tests on this platform as well if you want.

I don't want to remove the change in #11919 right away, since it currently still documents an issue with this platform. How should I document this? An issue?

kYc0o commented on cpu/cortexm_common/include/irq_arch.h:
```diff
@@ -18,41 +18,59 @@
+static inline __attribute__((always_inline)) unsigned int irq_disable(void)
 {
     uint32_t mask = __get_PRIMASK();
+
```
I guess this extra line can be removed.

fjmolinas replied:
Yep will do once #14011 is in

maribu replied:
ping

fjmolinas replied:
> ping

Sorry for the delay, Friday was a day off in France.

Regarding the extra line. I removed it just to remember why I had actually added it, when `uncrustifying ` that extra white line gets added, I think its better to leave it in because of that.

kYc0o commented on cpu/cortexm_common/include/irq_arch.h:
```diff
@@ -18,41 +18,59 @@
+    unsigned result = __get_PRIMASK();
     __enable_irq();
-    return __get_PRIMASK();
+    return result;
```
What's the actual difference here regarding the original code? By inlining the function it is the "extra" variable required? By looking at the sizes it's obviously not affecting, just a question out of curiosity.

maribu replied:
The return value here is never used in the whole code base of RIOT. There was a lengthy discussion about dropping it altogether that somehow never concluded to just dropping it. Maybe it is time again to revisit this PR and check if we now can just get rid of that.

But back to the size question: I think for the assembly it is just two instructions in different order now, as `__STATIC_FORCEINLINE __enable_irq(void)` leaves the register used to pass the return value untouched (according to the clobbers in the inline assembly). So the variable should be only present in C, but not in the machine code.

maribu replied:
I can't find the PR anymore :-/

kYc0o commented:
>I don't want to remove the change in #11919 right away, since it currently still documents an issue with this platform. How should I document this? An issue?

Yeah, an issue just to not forget to try this function back onto that platform. The already existing comments on the code are IMHO enough to explain the situation.

fjmolinas commented:
> Yeah, an issue just to not forget to try this function back onto that platform. The already existing comments on the code are IMHO enough to explain the situation.

Opened issue and removed special handling, rebased to get the #14011 changes.

maribu commented:
Code-wise, I'd ACK. Will the CI automagically pick up all tests that can be run using `make test` and run it? If so, no manual testing would be needed :-)

fjmolinas commented:
> Code-wise, I'd ACK. Will the CI automagically pick up all tests that can be run using make test and run it? If so, no manual testing would be needed :-)

I can do a run on many boards, no need to hurry this.

fjmolinas commented:
> Code-wise, I'd ACK. Will the CI automagically pick up all tests that can be run using make test and run it? If so, no manual testing would be needed :-)

It will run on `nrf52dk` and `samr21-xpro` at least, I'll run on more boards as well ust in case.

maribu commented:
Tested successfully using `tests/irq_disable_restore` on:

- Nucleo-F030R8 (STM32F030R8)
- Nucleo-F072RB (STM32F072RB)
- Nucleo-F303RE (STM32F303RE)
- Nucleo-F446RE (STM32F446RE)
- Nucleo-F767ZI (STM32F767ZI)
- Bluepill (STM32F103C8)
- micro:bit (nRF51)
- Arduino-Due (SAM3X8E)

maribu commented:
travis has some feedback as well

maribu commented:
> It will run on `nrf52dk` and `samr21-xpro` at least, I'll run on more boards as well ust in case.

Seems the CI tests failed for the nRF52. I don't have my nRF52 boards at hand to check if there is just an issue with the board used for the CI tests, or if this PR indeed causes regressions.

fjmolinas commented:
> Seems the CI tests failed for the nRF52. I don't have my nRF52 boards at hand to check if there is just an issue with the board used for the CI tests, or if this PR indeed causes regressions.

I'll run all the failing tests on iot-lab, although it seems to me like its an issue with the board, since there was  a lot of failures on last nightly as well.

fjmolinas commented:
Squashed travis comment right away.

fjmolinas commented:
Huh... it still says I need to uncrustify but when I do there is no diff..

fjmolinas commented:
<details><summary>TOOLCHAIN=llvm BUILD_IN_DOCKER=1  ./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . nrf52dk  --with-test-only --applications="tests/driver_nrfmin tests/cpp11_condition_variable tests/bench_xtimer_load tests/driver_apds99xx tests/bench_thread_yield_pingpong tests/pkg_tinycrypt tests/pkg_cn-cbor tests/irq tests/gnrc_sixlowpan tests/msg_try_receive tests/gnrc_sock_neterr tests/periph_gpio tests/pkg_cayenne-lpp tests/pthread_rwlock tests/xtimer_remove tests/pkg_qcbor tests/ztimer_msg tests/suit_manifest tests/pkg_tiny-asn1 tests/progress_bar tests/pthread tests/thread_flags_xtimer tests/thread_msg_block_race tests/xtimer_now64_continuity" -j3</summary>

```
#no failures
```

</details>


<details><summary>BUILD_IN_DOCKER=1 ./dist/tools/compile_and_test_for_board/compile_and_test_for_board.py . nrf52dk  --with-test-only --applications="tests/bench_runtime_coreapis tests/driver_grove_ledbar tests/bench_mutex_pingpong tests/isr_yield_higher tests/gnrc_netif tests/gnrc_sock_async_event tests/thread_flood tests/rmutex_cpp tests/pkg_monocypher tests/pkg_jsmn tests/pkg_libbase58 tests/pkg_tinycbor tests/struct_tm_utility tests/pthread_barrier tests/thread_msg tests/trickle tests/xtimer_mutex_lock_timeout tests/xtimer_usleep tests/cond_order tests/bloom_bytes" -j3  </summary>

```
#no failures
```

</details>

fjmolinas commented:
I think its safe to launch tests on the next compile once `uncrustify` issues are fixed all failing tests succeed locally so I'm blaming the worker https://github.com/RIOT-OS/RIOT/pull/13999#issuecomment-623972182

maribu commented:
Btw, when running

```
make BOARD=nucleo-f767zi -C tests/bench_mutex_pingpong flash term
``

With current `master` (`2d1d13d45276e238e7ac65fe80d4c54ddd8c65a8`):
```
2020-05-08 10:43:17,682 # Help: Press s to start test, r to print it is ready
s
2020-05-08 10:43:20,082 # START
2020-05-08 10:43:20,087 # main(): This is RIOT! (Version: 2020.07-devel-482-g2d1d1)
2020-05-08 10:43:20,088 # main starting
2020-05-08 10:43:21,090 # { "result" : 172111 }
2020-05-08 10:44:10,656 # Exiting Pyterm
```

And with this PR (rebased on top of `2d1d13d45276e238e7ac65fe80d4c54ddd8c65a8`):
```
2020-05-08 10:44:29,354 # READY
s
2020-05-08 10:44:30,137 # START
2020-05-08 10:44:30,144 # main(): This is RIOT! (Version: 2020.07-devel-486-ga43ecf-fjmolinas)
2020-05-08 10:44:30,145 # main starting
2020-05-08 10:44:31,147 # { "result" : 209912 }
```

Roughly 22% faster! :-)

fjmolinas commented:
> One comment forgotten to address (see above). Please squash right away.

@maribu I answered the comment above!

fjmolinas commented:
> Roughly 22% faster! :-)

Thas huge! In https://github.com/RIOT-OS/RIOT/pull/13825 we were concerned about a 1% drop, and here we improve 22%!

maribu commented:
> Thas huge! In #13825 we were concerned about a 1% drop, and here we improve 22%!

I think we could even get another gain of at least 5% if we optimized for the common case. E.g. something link:

```C
static inline void mutex_lock(mutex_t *mutex)
{
    unsigned irqstate = irq_disable();
    /* Tell compiler to expect the if condition to be mostly true */
    if (__builtin_expect((mutex->queue.next == NULL), 1)) {
        /* Common case: Mutex is unlocked */
        mutex->queue.next = MUTEX_LOCKED;
        irq_restore(irqstate);
        return 1;
    }
    mutex_handle_uncommon_case(mutex);
}
```

maribu commented:
OK. [This is the diff](https://github.com/RIOT-OS/RIOT/compare/62072b14a0205df49f6eb2cc4377fd2dabcb4466..7ea87756277f0175e5bf9563667f3be66a16084e) to the state I last checked. This changed only white space and renamed an include guard - nothing of syntactical meaning has changed since. Thus, my testing is still up to date.

I now also tested in addition to the nRF52840-DK. All works :-)

maribu commented:
I wouldn't mind a second review/ACK here, as this PR has a high impact.

fjmolinas commented:
> I wouldn't mind a second review/ACK here, as this PR has a high impact.

Yep I agree!

fjmolinas commented:
@kaspar030 or @bergzand might be able to take a look at this one as well?

kaspar030 commented on core/include/irq.h:
```diff
@@ -22,11 +22,14 @@
 extern "C" {
 #endif
 
+#ifdef LEGACY_IRQ_API
```
I'm a bit late, but would like to propose two changes:

1. use IRQ_API_INLINED instead of LEGACY_IRQ_API

That turns around the logic, so only adapted architectures need to be changed. Less surprising to anyone having an arch outside riot.
Also, "LEGACY_IRQ_API" works only once, until we want to improve this again. then we'd have "EVEN_MORE_LEGACY_IRQ_API" ... So, instead give the define a name that describes what it does.


2. I think in this case defining "MAYBE_INLINE" to `static inline __attribute__((always_inline))`, if IRQ_API_INLINED is active, and to an empty define if not, saves us from duplicate declarations and doxygen.

maribu replied:
>  Less surprising to anyone having an arch outside riot.

Let us please not worry about out of tree archs. Maintaining and deprecating RIOTs public APIs is already enough pain in the ass and a significant toll in the development process; but for official APIs this is certainly justified. Worrying about breaking out-of-tree code that touches RIOT's internals is IMO really off limits.

Before the next release we should be done with the conversion. And IMO that is the point we should get rid of `IRQ_API_INLINED` or `LEGACY_IRQ_API`. So next release, out-of-tree archs would have to follow suit. Or better: Being PR'ed to get upstream and benefit from being maintained by the community.

fjmolinas replied:
I personally don't mind either way, I was about to push the requested change. As I see it `LEGACY_IRQ_API` means touching more stuff now, but none at the end of the process, `IRQ_API_INLINED` means touching less stuff now, and all the platforms at the end. But if we actually change them all by the release `MAYBE_INLINE` would also be removed, having it as a place holder just reduces duplicated lines during migration. So IMO not much is gained or lost by any of the approaches.

fjmolinas replied:
I pushed the requested change, I can either drop those or squash depending on what you prefer @kaspar030 @maribu

maribu replied:
I'm fine with either :-)

fjmolinas replied:
Squashed.

fjmolinas commented:
Build failed because there was no space left on worker https://ci.riot-os.org/RIOT-OS/RIOT/13999/b5e4224a6fe2de4aca35d90af5e76d5b6a5bf519/output/compile/tests/mtd_mapper/stk3600:gnu.txt

fjmolinas commented:
All green again @kaspar030 !

fjmolinas commented:
Thanks for the review @kaspar030 @maribu!

maribu commented:
Thanks for the PR. This is huge improvement (at least without LTO) :-)
